scala-protoc-plugins/scala-akka/AkkaGrpcServicePrinter.scala

```diff
@@ -12,91 +12,105 @@ final class AkkaGrpcServicePrinter(
+          .add(s"def ${method.name}(")
           .indent
           .add(s"request: ${method.inputType.scalaType},")
-          .add(s"responseObserver: ${observer(method.outputType.scalaType)}")
+          .add(s"responseObserver: ${observer(responseType(method))}")
           .outdent
-          .add(s"): Unit = {")
+          .add("): Unit =")
           .indent
-          .add("if (closed.get()) {")
+          .add("synchronized {")
           .indent
-          .add("responseObserver.onError(closingError)")
+          .add("if (closed) {")
+          .indent
+          .add(
+            "responseObserver.onError(com.daml.grpc.adapter.server.akka.ServerAdapter.closingError())"
+          )
           .outdent
           .add("} else {")
           .indent
           .add(
-            "val sink = com.daml.grpc.adapter.server.akka.ServerAdapter.toSink(responseObserver)"
+            s"val killSwitch = com.daml.grpc.adapter.server.akka.ServerAdapter.registerStream(${method.name}Source(request), responseObserver)"
           )
-          .add(s"${method.name}Source(request).via(killSwitch.flow).runWith(sink)")
-          .add("()")
+          .add("killSwitches = killSwitch :: killSwitches")
           .outdent
           .add("}")
           .outdent
           .add("}")
-          .add(
-            s"protected def ${method.name}Source(request: ${method.inputType.scalaType}): akka.stream.scaladsl.Source[${method.outputType.scalaType}, akka.NotUsed]"
-          )
+          .outdent
           .newline
-      case StreamType.Bidirectional =>
-        p
+          .add(s"protected def ${method.name}Source(")
+          .indent
+          .add(s"request: ${method.inputType.scalaType}")
+          .outdent
+          .add(s"): akka.stream.scaladsl.Source[${responseType(method)}, akka.NotUsed]")
+          .newline
+      case StreamType.Bidirectional => p
     }
   }
 
   private def traitBody: PrinterEndo = {
     val endos: PrinterEndo = { p =>
-      p.call(service.methods.map(m => serviceMethodSignature(m)): _*)
+      p.newline
+        .call(service.methods.map(m => serviceMethodSignature(m)): _*)
     }
 
     p =>
-      p.add("protected implicit def esf: com.daml.grpc.adapter.ExecutionSequencerFactory")
+      p.indent
+        .add("protected implicit def esf: com.daml.grpc.adapter.ExecutionSequencerFactory")
         .add("protected implicit def mat: akka.stream.Materializer")
-        .call(closureUtils)
-        .newline
+        .call(closeMethod)
         .call(endos)
+        .outdent
   }
 
-  private def closureUtils: PrinterEndo = { p =>
-    p.newline
-      .add(s"protected val killSwitch = akka.stream.KillSwitches.shared($killSwitchName)")
-      .add("protected val closed = new java.util.concurrent.atomic.AtomicBoolean(false)")
+  private def closeMethod: PrinterEndo = { p =>
+    p
+      .add("@volatile private var closed = false")
+      .add("@volatile private var killSwitches = List.empty[akka.stream.KillSwitch]")
```

Comment:
It seems to me we only add here the KillSwitch-es, and never remove. :thinking: This is a problem, no?
Probably we would need to improve here to implement some concurrent map/set based approach (SharedKillSwitch for example uses TrieMap)